### PR TITLE
US041 - Authorisation System

### PR DESCRIPTION
**Pull request by LMBishop**
Corresponding backend PR: https://github.com/anthony-raybould/bham1-back-end/pull/9

This PR adds authorisation on the frontent. It checks locally first if the logged in user has the necessary role before attempting to query the API, and adds tokens to the Authorization header.

Screenshot 2023-09-20 at 10 19 10

Screenshot 2023-09-20 at 10 19 31

### CODE REVIEW

File: src/service/jobRoleService.ts [hunk 1]
```diff
@@ -2,15 +2,15 @@ import axios from "axios";
 import type { JobRole } from "../model/jobRole";
 
 export const jobRoleService = {
-    async getJobRoles(): Promise<JobRole[]> {
+    async getJobRoles(token?: string): Promise<JobRole[]> {
```
Comment (TomBillingtonUK) [inline src/service/jobRoleService.ts:5]: I would say a token isn't optional. If the endpoint requires the token we should ensure we have it before making the API call

File: src/service/jobRoleService.ts [hunk 2]
```diff
@@ -2,15 +2,15 @@ import axios from "axios";
...
+                headers: { Authorization: `Bearer ${token}` },
+            });
             if (response.status === 200) {
                 return response.data
             }
-        }
-        catch (e) {
-            throw new Error("Failed to get job roles")
-        }
+        } catch (e) { }
```
Comment (TomBillingtonUK) [inline src/service/jobRoleService.ts]: Did you mean to have an empty catch?

File: test/unit/controller/authControllerTest.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+// import chai from 'chai';
```
Comment (TomBillingtonUK) [inline test/unit/controller/authControllerTest.ts]: Did you not like these tests?

File: src/router.ts [hunk 4]
```diff
@@ -1,11 +1,20 @@
...
-import { JobRoles } from "./controller/jobRoleController";
 import { Auth } from "./controller/authController";
+import { JobRoles } from "./controller/jobRoleController";
+import { requireLoggedIn, requireLoggedOut, requireRole, user } from "./middleware/authorisation";
+
 const router = express.Router();
 
-router.get("/login", Auth.get)
-router.post("/login", Auth.post)
+router.use(user());
```
Comment (TomBillingtonUK) [inline src/router.ts]: I would put this in app.ts and leave the router to just handle routes

File: src/middleware/authorisation.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
+import { Request, Response, NextFunction } from "express";
+
+export function requireRole(requiredRole: string) {
```
Comment (TomBillingtonUK) [inline src/middleware/authorisation.ts:3]: You don't need to wrap the middleware in a function. You could have the inner function exported, and in the route you just use the function name without the brackets, and express will do the rest.

It may not work if you are passing in params but should for the other ones

File: src/middleware/authorisation.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    return function (req: Request, res: Response, next: NextFunction) {
+        if (req.session && (req.session.user?.role == 'Admin' || req.session.user?.role === requiredRole)) {
+            next();
+        } else {
```
Comment (TomBillingtonUK) [inline src/middleware/authorisation.ts:11]: You could remove the else and nested if by doing return next() and putting the logic underneath

File: src/middleware/authorisation.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import { Request, Response, NextFunction } from "express";
```
Comment (TomBillingtonUK) [inline src/middleware/authorisation.ts]: Missing unit tests for this file

File: src/middleware/authorisation.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            res.redirect('/');
+        } else {
+            next();
+        }
+    }
+}
+
+export function requireLoggedIn() {
+    return function (req: Request, res: Response, next: NextFunction) {
+        if (req.session && req.session.user) {
```
Comment (TomBillingtonUK) [inline src/middleware/authorisation.ts]: Is it worth checking the token is there as well?

File: test/unit/controller/jobRoleControllerTest.ts [hunk 9]
```diff
@@ -4,13 +4,23 @@ import express from 'express';
 import sinon from 'sinon';
```
Comment (TomBillingtonUK) [inline test/unit/controller/jobRoleControllerTest.ts:4]: I'm not sure what I feel about these unit tests as I wouldn't personally use them to test the output HTML. I would just test that the correct methods were called (res.render, res.redirect) with the expected parameters.

File: test/unit/service/authServiceTests.ts [hunk 10]
```diff
@@ -1,49 +1,88 @@
...
+    
+    it('should handle error when server error', async () => {
+      const mock = new MockAdapter(axios);
+      const token = 'token';
+      mock.onGet(`${process.env.API_URL}api/whoami`, { headers: { Authorization: `Bearer ${token}` } }).reply(500);
+        
+      try {
+        await authService.whoami(token);
+      } catch (error) {
+        expect(error.message).to.equal('Failed to fetch user');
```
Comment (TomBillingtonUK) [inline test/unit/service/authServiceTests.ts]: You should be able to use expect throws to check for exceptions and then you don't need the try/catch blocks

File: src/service/jobRoleService.ts (hunk 1 above)
Comment (LMBishop) [reply src/service/jobRoleService.ts:5]: I did that so it would compile since the token in the session could be undefined. Is there a better approach rather than just checking in every single controller whether the session exists, given each authorised route already has middleware to ensure it does?

File: src/service/jobRoleService.ts (hunk 2 above)
Comment (LMBishop) [reply src/service/jobRoleService.ts]: Yeah, otherwise I'd just duplicate line 14 inside the catch block and outside.

File: test/unit/controller/jobRoleControllerTest.ts (hunk 9 above)
Comment (LMBishop) [reply test/unit/controller/jobRoleControllerTest.ts:4]: I guess I can look into that, but I'll change the job roles stuff in a separate PR.

File: src/app.ts [hunk 11]
```diff
@@ -2,7 +2,9 @@ import expressSession from 'express-session';
 import express from 'express';
 import nunjucks from 'nunjucks';
 import path from 'path';
-import router from "./router"
+import router from './router';
+import { User } from './model/user';
```
Comment (TomBillingtonUK) [inline src/app.ts]: should be import type

File: src/middleware/authorisation.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
+import { Request, Response, NextFunction } from "express";
```
Comment (TomBillingtonUK) [inline src/middleware/authorisation.ts]: import type

File: src/service/authService.ts [hunk 13]
```diff
@@ -1,26 +1,38 @@
 import type { Login } from "../model/auth";
 import axios from "axios";
-import * as dotenv from 'dotenv';
+import { User } from "../model/user";
```
Comment (TomBillingtonUK) [inline src/service/authService.ts]: import type

File: src/service/jobRoleService.ts [hunk 14]
```diff
@@ -2,14 +2,14 @@ import axios from "axios";
 import type { JobRole } from "../model/jobRole";
 
 export const jobRoleService = {
-    async getJobRoles(): Promise<JobRole[]> {
+    async getJobRoles(token?: string): Promise<JobRole[]> {
+        if (!token) throw new Error("No token provided")
```
Comment (TomBillingtonUK) [inline src/service/jobRoleService.ts]: What is shown to the user on this path? I think this error floats all the way to the top of the service

File: test/unit/service/authServiceTest.ts [hunk 15]
```diff
@@ -0,0 +1,88 @@
+var chai = require('chai');
```
Comment (TomBillingtonUK) [inline test/unit/service/authServiceTest.ts]: Should be using typescript imports not require in the file

File: test/unit/service/jobRoleServiceTest.ts [hunk 16]
```diff
@@ -25,7 +25,7 @@ describe('jobRoleService', () => {
         }];
```
Comment (TomBillingtonUK) [inline test/unit/service/jobRoleServiceTest.ts:25]: You should add a test to cover the is token null path which you added in the PR

**Review by TomBillingtonUK — APPROVED** [review]

**Review by anthony-raybould-kainos — APPROVED** [review]
When eventually merging to main, change the UI tests to get their driver from the buildDriver function (currently in PR#13, but hopefully in main by the time this is through QA) to avoid them breaking in the GitHub action.
